content/js1/sprints/3/prep/index.md

```diff
@@ -8,4 +8,634 @@ backlog= 'Module-JS1'
+
+📓 We can and should continually check our assertions to see if our function’s current output meets our target output.
+
+So we have the following:
+
+```js {linenos=table,linenostart=1}
+function formatAs12HourClock(time) {
+  return `${time} am`;
+}
+
+const currentOutput = formatAs12HourClock("08:00");
+const targetOutput = "08:00 am";
+console.assert(
+  currentOutput === targetOutput,
+  "current output: %s, target output: %s",
+  currentOutput,
+  targetOutput
+);
+```
+
+✅ Nothing is printed to the console, so this assertion is working 😎
+
+### 💼 Checking different cases
+
+So far we’ve only created assertions that check the function’s behaviour for times _after midnight and before midday_. In these cases, there is a clear pattern: take the current time and add `"am"` to the end.
+
+However, we need to assert the function behaves correctly in cases when the time is later than midday. Let’s create an assertion for our function when passed an input of `"23:00"`:
+
+```js {linenos=table,linenostart=1,hl_lines=["15-21"]}
+function formatAs12HourClock(time) {
+  return `${time} am`;
+}
+
+const currentOutput = formatAs12HourClock("08:00");
+const targetOutput = "08:00 am";
+console.assert(
+  currentOutput === targetOutput,
+  "current output: %s, target output: %s",
+  currentOutput,
+  targetOutput
+);
+
+const currentOutput = formatAs12HourClock("23:00");
+const targetOutput = "11:00 pm";
+console.assert(
+  currentOutput === targetOutput,
+  "current output: %s, target output: %s",
+  currentOutput,
+  targetOutput
+);
+```
+
+### 🔄 Reusing variable names
+
+When we run the file with Node, we get an error in the console:
+
+```bash
+SyntaxError: Identifier 'currentOutput' has already been declared
+```
+
+In this case, identifier means name of a variable, so in a variable declaration like
+
+```js
+const currentOutput = formatAs12HourClock("08:23");
+```
+
+`currentOutput` is the **identifier**.
+
+When an error is thrown, it means the execution of the program stops at this point and an error report is sent to the user.
+However, we want to do multiple assertions whilst using the same variables.
+
+To reuse variable names, we can make use of a block declaration as follows:
+
+```js
+{
+  StatementList;
+}
+```
+
+🔑 A **block** is a region of code defined by a set of curly braces in which we write any number of statements.
+
+As with function scope, **block scope** means that variables declared inside a block are only accessible inside that block. This means we can declare a variable with the same name in 2 different blocks and we won't get a `SyntaxError`:
+
+```js title="problem.js"
+function formatAs12HourClock(time) {
+  return `${time} am`;
+}
+
+{
+  const currentOutput = formatAs12HourClock("08:00");
+  const targetOutput = "08:00 am";
+  console.assert(
+    currentOutput === targetOutput,
+    "current output: %s, target output: %s",
+    currentOutput,
+    targetOutput
+  );
+}
+
+{
+  // ❌ this assertion now fails
```

Comment:
I strongly agree that sneaking TDD in is great, and I agree that introducing "failing for the right reason" is good, but I'd lean pretty heavily towards doing so intentionally. Writing an intentionally failing test, or writing a surprisingly failing test are both great introductions to this, but confounding that with "We introduced a SyntaxError (and maybe the first SyntaxError we've ever encountered), oh no" is more likely to make people confused and scared, rather than appreciate the value of an intentionally failing test...

Reply:
Yeah that's fair 👍🏻